test/Npgsql.Tests/MultipleHostsTests.cs

```diff
@@ -513,6 +513,20 @@ public void Cluster_state_cache_basic()
             Assert.AreEqual(ClusterState.PrimaryReadWrite, ClusterStateCache.GetClusterState(host, 5432, ignoreExpiration: true));
         }
 
+        [Test]
+        public async Task ClusterUnknownStateOnConnectionFailure()
```

Comment:
Maybe add a test that simulates an IOException during query execution (with the mock)? It would also make sure that the pool is cleared.